File: Yogscart Alpha/Assets/Scripts/Kart Scripts/Racer_AI.js
#pragma strict

private var td : TrackData;
private var pf : Position_Finding;
private var ks : kartScript;
//private var ki : kartItem;
private var gd : CurrentGameData;

//0 - 50cc, 1 - 100cc, 2 - 150cc, 3 - Insane
var Stupidity : int; //Bigger the number, stupider the AI.

var angleRequired : float = 0f;
private var steering : int;

private var nTarget : Vector3;
private var targestPos : int = -1;

private var usedItem : boolean = false;
private var lastAngle : float;

private var adjusterFloat : float;

function Awake(){
td = GameObject.Find("Track Manager").GetComponent(TrackData);
pf = transform.GetComponent(Position_Finding);
ks = transform.GetComponent(kartScript);
//ki = transform.GetComponent(kartItem);
//gd = GameObject.Find("GameData").GetComponent(CurrentGameData);

adjusterFloat = Random.Range(-8f,8f);

}

function Update () {
//Calculate Item

/*if(ki.heldPowerUp != -1 && usedItem == false){

if(Stupidity < 5)
{

if(gd.PowerUps[ki.heldPowerUp].type == ItemType.UsableAsShield)
useShield();
else if(gd.PowerUps[ki.heldPowerUp].type == ItemType.Projectile)
FireProjectile();
else 
useItemRandom();

}else
useItemRandom();

usedItem = true;
}*/

//Calculate Steering
if(NumClamp(pf.currentPos+1,0,td.PositionPoints.Length) != targestPos ){

var nextChump : int;
if(pf.currentPos+1 < td.PositionPoints.Length)
nextChump = pf.currentPos+1;
else
nextChump = 0;

var Target : Vector3 = td.PositionPoints[nextChump].position;
var lastTarget : Vector3 = td.PositionPoints[pf.currentPos].position;

adjusterFloat += Random.Range(-3f,3f);
adjusterFloat = Mathf.Clamp(adjusterFloat,-6f,6f);

var Adjuster = Vector3.Cross((Target-lastTarget).normalized,transform.up) * adjusterFloat;

nTarget = Target + Adjuster;

targestPos = nextChump;

}

if(!ks.locked){
var NeededDirection : Vector3 = nTarget - transform.position;
NeededDirection.y = 0;

var fireDirection = transform.right;
fireDirection.y = 0;

var angle : float = Vector3.Angle(fireDirection,NeededDirection);

Debug.DrawRay(transform.position,transform.forward*5,Color.red);
Debug.DrawRay(transform.position,NeededDirection,Color.green);


if(angle > 90+angleRequired)
steering = -1;
else if(angle < 90-angleRequired)
steering = 1;
else{
steering = 0;
ks.drift = false;
}

if(steering != 0)
ks.throttle = 0.75;

//Calculate Throttle
if(angle >= 120 || angle  <= 60){

if(Stupidity < 6)
ks.drift = true;
ks.throttle = 0.5;

}else{
ks.throttle = 1;
}

if(ks.drift == true && lastAngle < angle)
ks.drift = false;

var testangle : float = Vector3.Angle(transform.forward,NeededDirection);
if(testangle >= 135){
steering = 1;
ks.throttle = 0.25;
}

var relativeVelocity : Vector3 = transform.InverseTransformDirection(GetComponent.<Rigidbody>().velocity);

if(ks.ExpectedSpeed > 1 && relativeVelocity.z < 1) //Presume something is blocking the kart.
reversing = true;

if(reversing){

Debug.DrawRay(transform.position,transform.forward*6,Color.green);
Debug.DrawRay(transform.position + transform.right,transform.forward*6,Color.green);
Debug.DrawRay(transform.position - transform.right,transform.forward*6,Color.green);

if(!Physics.Raycast(transform.position,transform.forward,6)&&!Physics.Raycast(transform.position + transform.right,transform.forward,6)&&!Physics.Raycast(transform.position - transform.right,transform.forward,6))
reversing = false;

steering = -steering;
ks.throttle = -1;
}

ks.steer = steering;

}

if(ks.startBoostVal != -1){
if(Stupidity < 4 && ks.startBoostVal <= 2)
ks.throttle = 1;

if(Stupidity > 8 && ks.startBoostVal <= 3)
ks.throttle = 1;
}

lastAngle = angle;

}

var reversing : boolean;

function useItemRandom(){
/*Debug.Log("Started Iteming!");
yield WaitForSeconds(Random.Range(5,20));
ki.input = true;
yield;
yield;
Debug.Log("Done Iteming!");
ki.input = false;
usedItem = false;*/
}

function useShield(){
/*Debug.Log("Started Iteming!");
ki.input = true;
while(ki.heldPowerUp != -1)
yield;
Debug.Log("Done Iteming!");
ki.input = false;
usedItem = false;*/
}

function FireProjectile()
{
/*
while(true)
{
var hit : RaycastHit;

if(Physics.Raycast(transform.position,transform.forward,hit,25) && hit.transform.tag == "Kart")
{
ki.input = true;
break;
}

yield;
}

ki.input = false;
usedItem = false;
*/
}

function NumClamp(val : int,min : int,max : int){

while(val > max-1)
val -= (max-min);

while(val < min)
val += (max-min);


return val;
}

File: Yogscart C#/Assets/Scripts/Kart Scripts/RacerAI.cs
using UnityEngine;
using System.Collections;

public class RacerAI : MonoBehaviour
{

    //Straight Port of JS Version
    private kartScript ks;
    private TrackData td;
    private PositionFinding pf;

    public int stupidity = -1; //Bigger the number, stupider the AI.

    private float angleRequired = 3f;
    private float turnSpeed = 15f;
    private float turnAngle = 30f;

    private float reverseDistance = 10f;

    private Transform turnPoint;
    private int turnPointInt;

    private bool reversing = false, startDrive = true;

    public float adjusterFloat = -999f;

    // Use this for initialization
    void Awake()
    {
        ks = GetComponent<kartScript>();
        td = GameObject.FindObjectOfType<TrackData>();
        pf = GetComponent<PositionFinding>();

        if (ks.locked)
            startDrive = false;
    }

    // Update is called once per frame
    void FixedUpdate()
    {
        //Wait till positions are set to create Adjuster
        if (adjusterFloat <= -999 && pf.position != -1)
        {
            adjusterFloat = 5 - ((pf.position % 3) * 5);
        }

        int currentPos = pf.currentPos;
        Vector3 transformPos = transform.position;
        transformPos.y = 0;

        int pointTotal = td.positionPoints.Count;

        Vector3 nextPoint = td.positionPoints[MathHelper.NumClamp(currentPos + 1, 0, pointTotal)].position;
        nextPoint.y = 0;

        Vector3 nextNextPoint = td.positionPoints[MathHelper.NumClamp(currentPos + 2, 0, pointTotal)].position;
        nextNextPoint.y = 0;

        Vector3 currentPoint = td.positionPoints[currentPos].position;
        currentPoint.y = 0;

        Vector3 desiredDirection = nextPoint - currentPoint;
        desiredDirection.y = 0;

        Vector3 fireDirection = transform.right;
        fireDirection.y = 0;

        float angle = Vector3.Angle(fireDirection, desiredDirection);
        int turnRequired = CheckAngle(angle);

        if (startDrive)
            ks.throttle = 1f;
        else
            ks.throttle = 0f;

        if (turnPointInt == currentPos)
        {
            turnPoint = null;
        }

        if (turnPoint == null)
        {

            turnPoint = td.positionPoints[MathHelper.NumClamp(currentPos + 1, 0, pointTotal)];
            turnPointInt = MathHelper.NumClamp(currentPos + 1, 0, pointTotal);

            if (adjusterFloat > -999)
            {
                adjusterFloat += Random.Range(-1f, 1f);
                float limit = turnPoint.GetComponent<PointHandler>().roadWidth;
                adjusterFloat = Mathf.Clamp(adjusterFloat, -limit, limit);
            }

        }
        else
        {
            var tpDistance = Vector3.Distance(transformPos, turnPoint.position);

            //If turn within turn range start to turn
            if (tpDistance < 7.5f && !reversing)
            {
                SlowDownCar(Vector3.Angle(transform.forward, desiredDirection));
                Debug.DrawRay(transform.position, desiredDirection, Color.green);
            }
            else
            {
                //If not within range of next point, aim kart towards the point	
                var Adjuster = Vector3.Cross((turnPoint.position - currentPoint).normalized, transform.up) * adjusterFloat;
                Debug.DrawLine(transform.position, turnPoint.position + Adjuster, Color.green);

                Vector3 NeededDirection = (turnPoint.position + Adjuster) - transformPos;
                NeededDirection.y = 0;

                float nAngle = Vector3.Angle(fireDirection, NeededDirection);
                int nTurnRequired = CheckAngle(nAngle);

                turnRequired = nTurnRequired;
                SlowDownCar(Vector3.Angle(transform.forward, NeededDirection));
            }
        }

        //Reverse if Kart hits somethings
        if (ks.ExpectedSpeed > ks.maxSpeed/2f && Mathf.Abs(ks.ActualSpeed) < 1) //Presume something is blocking the kart.
            reversing = true;

        if (reversing)
        {
            turnRequired = (int)Mathf.Sign(-turnRequired);
            ks.throttle = -1;

            var checkPos = transform.position + Vector3.up;

            if (!Physics.Raycast(checkPos, transform.forward, reverseDistance) && !Physics.Raycast(checkPos + transform.right, transform.forward, reverseDistance) && !Physics.Raycast(checkPos - transform.right, transform.forward, reverseDistance))
                reversing = false;
        }

        //Handles Start Boosting
        if (kartScript.startBoostVal != -1)
        {
            if ((stupidity < 4 && kartScript.startBoostVal <= 2) || (stupidity > 8 && kartScript.startBoostVal <= 3))
                ks.throttle = 1;

            if (kartScript.startBoostVal <= 1)
                startDrive = true;

        }

        ks.steer = turnRequired;

    }

    void SlowDownCar(float angle)
    {
        if (angle > turnAngle)
        {
            //Slow down to make turn
            if (stupidity < 5f)
            {
                if (ks.ActualSpeed >= turnSpeed)
                    ks.throttle = 0;
            }
            else
            {
                ks.drift = true;
            }
        }
        else
        {
            ks.drift = false;
        }
    }

    int CheckAngle(float angle)
    {
        if (angle > 90 + angleRequired)
            return -1;
        else if (angle < 90 - angleRequired)
            return 1;
        else
            return 0;
    }

}
